src/sctools/metrics/basemetric.py

```diff
@@ -0,0 +1,16 @@
+import abc
+
+
+class BaseMetric:
+
+    @abc.abstractmethod
+    def initialize(self):
+        """Do some kind of one time initialization for the metrics class"""
+
+    @abc.abstractmethod
+    def gather_metric(self, sam_record):
+        """What the metric to do with each read of a bam"""
+
+    @abc.abstractmethod
+    def calculate_and_output(self):
```

Comment:
This probably needs some parameters (like a file handle) -- that said I have some comments about this below to read first.